### ISSUE TITLE
Broadcasting matmul is much slower than corresponding einsum

### ISSUE THREAD

**Issue by francois-rozet** [pytorch/pytorch#110858]
### 🐛 Describe the bug

Asymmetric broadcasting `matmul` is much slower (hundreds of times slower) than the corresponding `einsum` operation, at least on CPU (I haven't tried on GPU yet).

```python
>>> import torch
>>> A = torch.randn(3, 64, 64)
>>> B = torch.randn(4096, 3, 64, 1) 
>>> %timeit A @ B
29.8 ms ± 697 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
>>> %timeit torch.einsum('ijk,...ikl->...ijl', A, B)
396 µs ± 2.84 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

### Versions

```
Collecting environment information...
PyTorch version: 2.0.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 23.04 (x86_64)
GCC version: (Ubuntu 12.3.0-1ubuntu1~23.04) 12.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.37

Python version: 3.9.16 | packaged by conda-forge | (main, Feb  1 2023, 21:39:03)  [GCC 11.3.0] (64-bit runtime)
Python platform: Linux-5.19.0-29-generic-x86_64-with-glibc2.37
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   48 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          16
On-line CPU(s) list:             0-15
Vendor ID:                       AuthenticAMD
Model name:                      AMD Ryzen 7 PRO 6850U with Radeon Graphics
CPU family:                      25
Model:                           68
Thread(s) per core:              2
Core(s) per socket:              8
Socket(s):                       1
Stepping:                        1
Frequency boost:                 enabled
CPU(s) scaling MHz:              44%
CPU max MHz:                     4768,0000
CPU min MHz:                     400,0000
BogoMIPS:                        5390,25
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca
Virtualization:                  AMD-V
L1d cache:                       256 KiB (8 instances)
L1i cache:                       256 KiB (8 instances)
L2 cache:                        4 MiB (8 instances)
L3 cache:                        16 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-15
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.25.1
[pip3] torch==2.0.1
[conda] Could not collect
```

**Comment by mingfeima** [comment]
@jiayisunx could you please take a look at this one? bmm has a fallback path when MNK < 400, not sure if this is the root cause.

**Comment by lezcano** [comment]
FWIW, the issue is also present on gpu

**Comment by hchau630** [comment]
Any updates on this? I'm on torch version 2.3.0 and I'm still seeing this issue. On an A40 GPU I see that `einsum` is 10x faster than `matmul` on these broadcasted matrix multiplications, which is unacceptably slow considering that `matmul` is used everywhere by everyone.

**Comment by lezcano** [comment]
This is still the case in eager. If you want speed, for this kind of shapes, you can use `torch.compile`. If you `torch.compile` `a @ b`, my benchmarks show that this is 2x faster than einsum.

**Comment by hchau630** [comment]
couldn't this be fixed just be calling `einsum` inside `matmul` for these shape?

**Comment by francois-rozet** [comment]
Maybe the CPU tag should be removed?

**Comment by lezcano** [comment]
For which shapes? What heuristic are you proposing concretely?

**Comment by lezcano** [comment]
Note that we can do this sort of things easily in `torch.compile`, as we can compare different configurations easily, and we can create efficient fused kernels when we need to, but on eager you have to do the dispatch statically, so you have to come up with complex heuristics, that may work on some hardware and might regress in other, so it's just trickier.

**Comment by hchau630** [comment]
Here's a simple suggestion: suppose we are computing `A @ B` with `A.shape == (m, n)`, `B.shape == (b, n, k)`. Then simply call `einsum` instead of `matmul` whenever `b > k`. More generally, let `b = max(B.size(d) for d in broadcasted_dims`, where `broadcasted_dims` are the dimensions of `B` which are broadcasted, and again we call `einsum` instead of `matmul` whenever `b > k`. My preliminary testing shows that this approach is almost always faster than `matmul`, and the performance improvement can be up to 50x.

Here are some plots illustrating the performance improvement. I compared the timing of this approach, which I call `einsum_matmul`, with a naive `matmul`, for tensors `A, B` with sizes `(n, n)` and `(b, n, k)` for various values of `b, n, k`. The logarithm (base 10) of the timing of `einsum_matmul`/`matmul` is plotted, so a negative value indicates a performance improvement. The first plot is on CPU on an M1 Macbook Pro, and the second plot is on an A40 GPU on a linux machine.
![timing_macos_cpu](https://github.com/pytorch/pytorch/assets/25211640/e3be095a-0f39-47a3-90ff-5d23e706ed7d)
![timing_linux_gpu](https://github.com/pytorch/pytorch/assets/25211640/97a59e2d-a686-423f-a648-582ca5e13005)

**Comment by hchau630** [comment]
Interestingly, the performance improvement is much smaller or even sometimes slower when gradient is required. Here are the corresponding plots where `A` and `B` both requires gradient.

![timing_macos_cpu_grad](https://github.com/pytorch/pytorch/assets/25211640/74096bbc-a8c3-435b-ad23-2544144e67b1)
![timing_linux_gpu_grad](https://github.com/pytorch/pytorch/assets/25211640/e5d72477-25de-43cb-8325-cdbad4b9486a)

**Comment by hchau630** [comment]
Actually this seems to be the symptom of a different issue. Based on my last point I looked at the timing between `A @ B` without gradient and `A @ B` with gradient. It seems that for these asymmetric batching shapes, `A @ B` without gradient is much slower than `A @ B` with gradient, and that this difference entirely accounts for the performance difference between `matmul` and `einsum`. Surely this is a bug?

**Comment by lezcano** [comment]
Nah, it's https://github.com/pytorch/pytorch/pull/76828/. If you want to test it properly with `requries_grad=True` you should also benchmark the gradients. See the `should_fold` function.

Also, to add a proper benchmark, we would need a comprehensive analysis on, say, an intel CPU with MKL enabled (fast path) and a GPU, better if it's an A100, for example. We would need to test a grid of matrices for `bxmxnxk` in the range 0...8192, for example. Powers of two and close to powers of two are important for vectorisation and so on.

**Comment by francois-rozet** [comment]
I reproduced this issue again, on GPU (A100), with the latest pytorch version (2.7.1). Could we remove the `module: cpu` label?

![Image](https://github.com/user-attachments/assets/f17548ed-6827-4ec2-a442-94913ec71f7d)

**Comment by hchau630** [comment]
I feel like this should be a high priority issue given that matmul is orders of magnitudes slower than einsum in a lot of cases. There are definitely tons of users being severely affected by this issue without knowing about it.
